ports/servoshell/app.rs

```diff
@@ -396,9 +414,17 @@ impl App {
         let mut embedder_messages = self.servo.as_mut().unwrap().get_events();
         let mut need_resize = false;
         let mut need_present = false;
+        let mut history_changed = false;
         loop {
             // Consume and handle those embedder messages.
-            need_present |= browser.handle_servo_events(embedder_messages);
+            let (new_need_present, new_history_changed) =
+                browser.handle_servo_events(embedder_messages);
+            need_present |= new_need_present;
+            history_changed |= new_history_changed;
+
+            if history_changed {
+                return Some(PumpResult::HistoryChanged);
```

Comment:
i checked the values returned by handle_servo_events, and sometimes both flags are true. when that happens, we won’t present the frame, which could make the compositor hang. what do you think about reworking PumpResult to allow for this? i think the cases should be something like

- Shutdown
- Continue { history_changed: bool, Present } where Present can be
    - Immediate
    - Deferred
    - None

Reply:
done in [7b74ee5](https://github.com/servo/servo/pull/30518/commits/7b74ee5350cd67d25ccc8489d7864c9fbd0b1f9e)